File: TrainPlugin/trainplugin/metrics.py
"""无依赖的系统指标采集（Linux /proc + 标准库）。

返回四项使用占比（百分比，0~100 或 0~100+）：
CPU、内存、磁盘、磁盘IO（iostat 式 %util）。
"""

import shutil
import time

_DEFAULT_INTERVAL = 0.2


def _cpu_times():
    with open("/proc/stat") as f:
        parts = f.readline().split()
    vals = [int(x) for x in parts[1:8]]  # user nice system idle iowait irq softirq steal
    total = sum(vals)
    idle = vals[3] + vals[4]  # idle + iowait
    return total, idle


def _mem_percent():
    info = {}
    with open("/proc/meminfo") as f:
        for line in f:
            key, _, value = line.partition(":")
            info[key.strip()] = int(value.split()[0])
    return (1 - info["MemAvailable"] / info["MemTotal"]) * 100


def _disk_percent():
    usage = shutil.disk_usage("/")
    return (1 - usage.free / usage.total) * 100


def _root_device():
    with open("/proc/mounts") as f:
        for line in f:
            parts = line.split()
            if len(parts) >= 2 and parts[1] == "/":
                return parts[0].rsplit("/", 1)[-1]
    return ""


def _disk_io_ms():
    dev = _root_device()
    if not dev:
        return 0.0
    with open("/proc/diskstats") as f:
        for line in f:
            parts = line.split()
            if len(parts) >= 13 and parts[2] == dev:
                return int(parts[12])  # time_doing_io（毫秒）
    return 0.0


def collect(interval=_DEFAULT_INTERVAL):
    """采集一次指标，返回 (cpu, memory, disk, disk_io) 四个百分比。"""
    try:
        total0, idle0 = _cpu_times()
        io0 = _disk_io_ms()
        time.sleep(interval)
        total1, idle1 = _cpu_times()
        io1 = _disk_io_ms()

        cpu = (1 - (idle1 - idle0) / (total1 - total0)) * 100 if total1 > total0 else 0.0
        mem = _mem_percent()
        disk = _disk_percent()
        diskio = (io1 - io0) / (interval * 1000) * 100
        return cpu, mem, disk, diskio
    except Exception:
        return 0.0, 0.0, 0.0, 0.0

File: DatabasePlugin/pkg/metrics.go
package pkg

import (
	"bufio"
	"os"
	"strconv"
	"strings"
	"time"
)

const metricsInterval = 200 * time.Millisecond

func readCPUTimes() (total, idle uint64, err error) {
	f, err := os.Open("/proc/stat")
	if err != nil {
		return 0, 0, err
	}
	defer f.Close()
	sc := bufio.NewScanner(f)
	if !sc.Scan() {
		return 0, 0, sc.Err()
	}
	parts := strings.Fields(sc.Text())
	if len(parts) < 8 {
		return 0, 0, os.ErrInvalid
	}
	var vals [7]uint64
	for i := 0; i < 7; i++ {
		v, err := strconv.ParseUint(parts[i+1], 10, 64)
		if err != nil {
			return 0, 0, err
		}
		vals[i] = v
	}
	for _, v := range vals {
		total += v
	}
	idle = vals[3] + vals[4]
	return total, idle, nil
}

func readMemPercent() (float32, error) {
	f, err := os.Open("/proc/meminfo")
	if err != nil {
		return 0, err
	}
	defer f.Close()
	info := make(map[string]uint64)
	sc := bufio.NewScanner(f)
	for sc.Scan() {
		kv := strings.SplitN(sc.Text(), ":", 2)
		if len(kv) != 2 {
			continue
		}
		key := strings.TrimSpace(kv[0])
		fields := strings.Fields(kv[1])
		if len(fields) == 0 {
			continue
		}
		n, err := strconv.ParseUint(fields[0], 10, 64)
		if err != nil {
			continue
		}
		info[key] = n
	}
	total := info["MemTotal"]
	avail := info["MemAvailable"]
	if total == 0 {
		return 0, os.ErrInvalid
	}
	return float32(1-float64(avail)/float64(total)) * 100, nil
}

func rootDevice() string {
	f, err := os.Open("/proc/mounts")
	if err != nil {
		return ""
	}
	defer f.Close()
	sc := bufio.NewScanner(f)
	for sc.Scan() {
		fields := strings.Fields(sc.Text())
		if len(fields) >= 2 && fields[1] == "/" {
			idx := strings.LastIndex(fields[0], "/")
			return fields[0][idx+1:]
		}
	}
	return ""
}

func readDiskIOMs() (float64, error) {
	dev := rootDevice()
	if dev == "" {
		return 0, os.ErrNotExist
	}
	f, err := os.Open("/proc/diskstats")
	if err != nil {
		return 0, err
	}
	defer f.Close()
	sc := bufio.NewScanner(f)
	for sc.Scan() {
		parts := strings.Fields(sc.Text())
		if len(parts) >= 13 && parts[2] == dev {
			return strconv.ParseFloat(parts[12], 64)
		}
	}
	return 0, os.ErrNotExist
}

// CollectMetrics 采集 CPU、内存、磁盘、磁盘IO 使用占比（百分比）。
func CollectMetrics() (cpu, memory, disk, diskIO float32) {
	t0, i0, err0 := readCPUTimes()
	io0, _ := readDiskIOMs()
	time.Sleep(metricsInterval)
	t1, i1, _ := readCPUTimes()
	io1, _ := readDiskIOMs()

	if err0 == nil && t1 > t0 {
		cpu = float32(1-float64(i1-i0)/float64(t1-t0)) * 100
	}
	memory, _ = readMemPercent()
	disk, _ = readDiskPercent()
	diskIO = float32((io1 - io0) / (metricsInterval.Seconds() * 1000) * 100)
	return
}
